File: 1-two-sum/1-two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // if the array length is less than 2, we don't have enough 2 indices to return
        if (nums.length < 2) {
            return null;
        }
        
        //this map holds the numbers that add to the target and their indices
        Map<Integer, Integer> map = new HashMap<>();
        
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(nums[i])) {
                return new int[]{Math.min(i, map.get(nums[i])), Math.max(i, map.get(nums[i]))};
            }
            map.put(target - nums[i], i);
        }
        return new int[]{}; //cannot find any pairs
    }
}

File: 1-two-sum/1-two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        map = {}
        for i in range(0, len(nums)):
            if nums[i] in map:
                return [min(i, map[nums[i]]), max(i, map[nums[i]])]
            else:
                map[target - nums[i]] = i
        return []